src/main/scala/async/Listener.scala

```diff
@@ -1,73 +1,171 @@
+  def apply[T](consumer: T => Unit): Listener[T] = acceptingListener(consumer)
+
+  /** A special class of listener that forwards the inner listener through the given source.
+    * For purposes of [[Async.Source.dropListener]] these listeners are compared for equality
+    * by the hash of the source and the inner listener.
+    */
+  abstract case class ForwardingListener[T](src: Async.Source[?], inner: Listener[T]) extends Listener[T]
+
+  object ForwardingListener:
+    /** Create an empty [[ForwardingListener]] for equality comparison. */
+    def empty[T](src: Async.Source[?], inner: Listener[T]) = new ForwardingListener(src, inner):
+      val lock = null
+      override def complete(data: T, source: Async.Source[T]) = ???
+      override def release(to: Listener.LockMarker) = ???
 
+
+  /** A helper instance that provides an uniquely numbered mutex. */
+  trait NumberedLock:
+    import NumberedLock._
+
+    val number = listenerNumber.getAndIncrement()
+    private val lock0 = ReentrantLock()
+
+    protected def acquireLock() = lock0.lock()
+    protected def releaseLock() = lock0.unlock()
+  object NumberedLock:
+    private val listenerNumber = java.util.concurrent.atomic.AtomicLong()
+
+/** A listener, representing an one-time value receiver of an `Async.Source`.
+  *
+  * Most of the time listeners should involve only calling a receiver function,
+  * and can be created by `Listener.apply` or `Listener.acceptingListener`.
+  *
+  * However, should the listener want to attempt synchronization, it has to
+  * expose some locking-related interfaces. See `lock` and `release`.
+  */
 trait Listener[-T]:
-  val topLock: Listener.TopLock | Null
+  import Listener._
+  /** Represents the exposed API for synchronization on listeners at receiving time.
+    * If the listener does not have any form of synchronization, [[lock]] should be `null`.
+    */
+  val lock: Listener.ListenerLock | Null
 
+  /** Complete the listener with the given item, from the given source.
+    * **If the listener exposes a [[Listener.ListenerLock]]**, it is required to acquire this lock completely
+    * (either through [[lockCompletely]] or through manual locking of every layer)
+    * before calling [[complete]].
+    * For performance reasons, this condition is usually not checked and will end up
+    * causing unexpected behavior.
+    *
+    * The listener should automatically release the lock of itself and any underlying listeners,
+    * however this usually is done automatically by calling the inner listener's [[complete]]
+    * recursively.
+    */
   def complete(data: T, source: Async.Source[T]): Unit
-  def release(to: Listener.ReleaseBoundary): Listener[?] | Null
+  /** Release the current lock without resolving the listener with any items, if the
+    * current listener is before or the same as the current [[Listener.LockMarker]].
+    */
+  protected def release(to: Listener.LockMarker): Listener[?] | Null
```

Comment:
Uhm, why did we introduce this on the Listener instead of the ListenerLock?

Reply:
mmm, probably a better idea to move it to ListenerLock